clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -128,6 +128,170 @@ static Attr *handleIntelFPGALoopAttr(Sema &S, Stmt *St, const ParsedAttr &A) {
 }
 
+static bool checkSYCLIntelFPGAIVDepSafeLen(Sema &S, llvm::APSInt &Value,
+                                           Expr *E) {
+  if (!Value.isStrictlyPositive())
+    return S.Diag(E->getExprLoc(),
+                  diag::err_attribute_requires_positive_integer)
+           << "'ivdep'" << /* positive */ 0;
+  return false;
+}
+
+enum class IVDepExprResult {
+  Invalid,
+  Null,
+  Dependent,
+  Array,
+  SafeLen,
+};
+
+static IVDepExprResult HandleFPGAIVDepAttrExpr(Sema &S, Expr *E,
+                                               unsigned &SafelenValue) {
+  if (!E)
+    return IVDepExprResult::Null;
+
+  if (E->isInstantiationDependent())
+    return IVDepExprResult::Dependent;
+
+  llvm::APSInt ArgVal;
+
+  if (E->isIntegerConstantExpr(ArgVal, S.getASTContext())) {
+    if (checkSYCLIntelFPGAIVDepSafeLen(S, ArgVal, E))
+      return IVDepExprResult::Invalid;
+    SafelenValue = ArgVal.getZExtValue();
+    return IVDepExprResult::SafeLen;
+  }
+
+  if (isa<DeclRefExpr>(E)) {
+    if (!cast<DeclRefExpr>(E)->getType()->isArrayType()) {
+      S.Diag(E->getExprLoc(), diag::err_ivdep_declrefexpr_arg);
+      return IVDepExprResult::Invalid;
+    }
+    return IVDepExprResult::Array;
+  }
+
+  S.Diag(E->getExprLoc(), diag::err_ivdep_unknown_arg);
+  return IVDepExprResult::Invalid;
+}
+
+// Note: At the time of this call, we don't know the order of the expressions,
+// so we name them vaguely until we can figure it out.
+SYCLIntelFPGAIVDepAttr *
+Sema::BuildSYCLIntelFPGAIVDepAttr(const AttributeCommonInfo &CI, Expr *Expr1,
+                                  Expr *Expr2) {
+  unsigned SafelenValue = 0;
+  IVDepExprResult E1 = HandleFPGAIVDepAttrExpr(*this, Expr1, SafelenValue);
+  IVDepExprResult E2 = HandleFPGAIVDepAttrExpr(*this, Expr2, SafelenValue);
+
+  if (E1 == IVDepExprResult::Invalid || E2 == IVDepExprResult::Invalid)
+    return nullptr;
+
+  if (E1 == E2 && E1 != IVDepExprResult::Dependent &&
+      E1 != IVDepExprResult::Null) {
+    Diag(Expr2->getExprLoc(), diag::err_ivdep_duplicate_arg);
+    return nullptr;
+  }
+
+  // Try to put Safelen in the 1st one so codegen can count on the ordering.
+  Expr *SafeLenExpr;
+  Expr *ArrayExpr;
+  if (E1 == IVDepExprResult::SafeLen) {
+    SafeLenExpr = Expr1;
+    ArrayExpr = Expr2;
+  } else {
+    SafeLenExpr = Expr2;
+    ArrayExpr = Expr1;
+  }
+
+  return new (Context)
+      SYCLIntelFPGAIVDepAttr(Context, CI, SafeLenExpr, ArrayExpr, SafelenValue);
+}
+
+// Filters out any attributes from the list that are either not the specified
+// type, or whose function isDependent returns true.
+template <typename T>
+static void FilterAttributeList(ArrayRef<const Attr *> Attrs,
+                    SmallVectorImpl<const T *> &FilteredAttrs) {
+
+  llvm::transform(Attrs, std::back_inserter(FilteredAttrs), [](const Attr *A) {
+    if (const auto *Cast = dyn_cast_or_null<const T>(A))
+      return Cast->isDependent() ? nullptr : Cast;
```

Comment:
I kinda cringe at the thought that someone might attempt to use it for a `isDependent`-non-implementing subclass. Adding a check for `HasCustomTypeTransform` might alleviate this *some*, but this is too shallow a guarantee (besides, one might attempt implementing `isDependent` without setting `HasCustomTypeTransform = 1`.

Also, the need to implement `isDependent` is explicitly commented, so if there are no objections on @MrSidims' side, let's leave it this way.